### PR TITLE
Enable setting up paypal in paymentsheet

### PR DESCRIPTION
**Pull request by yuki-stripe**
## Summary
Allows Paypal to be used for PI+SFU and SI in PaymentSheet, but *don't allow* it to be displayed as a saved PM.

Includes a refactor that removes `supportsSaveAndReuse` and replaces with explicit `supportsSaveForFutureUseCheckbox` and `supportsSavedPaymentMethod`.  `supportsSaveAndReuse` controlled PMs availability for PI+SFU/SI **as well as** the 'save for future use' checkbox and its display as a saved PMs .  Paypal breaks this - it needs to be enabled for PI+SFU/SI but not the others.

## Motivation

- https://jira.corp.stripe.com/browse/RUN_MOBILESDK-1859
- For decision to enable support while hiding from saved payment methods list, see https://docs.google.com/document/d/1_bCPJXxhV4Kdgy7LX7HPwpZfElN3a2DcYUooiWC9SgM/edit#

## Testing
Manually tested these 3 flows, with a Customer so I can see saved PMs.

**PI**
image
✅ Successfully confirms
✅ Doesn't show "Save for future payments" checkbox

**PI+sfu**
image
✅ Successfully confirms
✅ Never appears in saved PMs list

**SI**
image

✅ Successfully confirms
✅ Never appears in saved PMs list

## Changelog
See CHANGELOG.md

### CODE REVIEW

File: StripePaymentSheet/StripePaymentSheet/Source/Categories/String+Localized.swift [hunk 1]
```diff
@@ -155,4 +155,32 @@ extension String.Localized {
...
+            "By confirming your payment with PayPal, you allow %@ to charge your PayPal account for future payments in accordance with their terms.",
+            "Paypal mandate text"
+        )
+    }
+
+    static var paypal_mandate_text_setup: String {
+        STPLocalizedString(
+            "By continuing to PayPal, you allow %@ to charge your PayPal account for future payments in accordance with their terms.",
+            "Paypal mandate text"
+        )
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/Categories/String+Localized.swift:184]: cc @jaynewstrom-stripe

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Views/SimpleMandateTextView.swift [hunk 2]
```diff
@@ -1,16 +1,17 @@
 //
-//  CashAppMandateView.swift
+//  SimpleMandateTextView.swift
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Views/SimpleMandateTextView.swift:2]: FYI @porter-stripe I noticed CashAppMandateView and SepaMandateView were identical, so I combined into `SimpleMandateTextView.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentSheet+API.swift [hunk 3]
```diff
@@ -101,26 +113,42 @@ extension PaymentSheet {
...
+                        with: authenticationContext,
+                        completion: paymentHandlerCompletion
+                    )
                 }
             // MARK: ↪ SetupIntent
             case .setupIntent(let setupIntent):
                 let setupIntentParams = confirmParams.makeParams(setupIntentClientSecret: setupIntent.clientSecret)
                 setupIntentParams.returnURL = configuration.returnURL
+                // Paypal requires mandate_data if setting up
+                if confirmParams.paymentMethodType.stpPaymentMethodType == .payPal {
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentSheet+API.swift:133]: Same here

**Review by porter-stripe — DISMISSED** [review]
LGTM, minor comment/questions

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,4 +1,6 @@
 ## x.x.x yyyy-mm-dd
+### PaymentSheet
+* [Added] Adds support for setting up Paypal using a SetupIntent or a PaymentIntent w/ setup_future_usage=off_session. Note: Paypal is in beta.
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply CHANGELOG.md:3]: Thank you!

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentSheet+API.swift [hunk 5]
```diff
@@ -101,26 +113,42 @@ extension PaymentSheet {
                     )
                     paymentIntentParams.returnURL = configuration.returnURL
                     paymentIntentParams.shipping = makeShippingParams(for: paymentIntent, configuration: configuration)
-                    paymentHandler.confirmPayment(paymentIntentParams,
-                                                  with: authenticationContext,
-                                                  completion: paymentHandlerCompletion)
+                    // Paypal requires mandate_data if setting up
+                    if confirmParams.paymentMethodType.stpPaymentMethodType == .payPal
+                        && paymentIntent.setupFutureUsage == .offSession
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentSheet+API.swift:118]: At this point we're making decisions around allowing setup per PM, can revisit this in the future

**Pull request merged**